### PR TITLE
ENH: stats: Add Skewed Cauchy Distribution

### LINKED ISSUE #7437: ENH: add skewed Cauchy distribution to stats
I recently had a need for the skewed Cauchy distribution described here:
https://en.wikipedia.org/wiki/Skewed_generalized_t_distribution#skewed_Cauchy_distribution
Especially important to me is the fit() method for this distribution.
I don't have the skill to implement this, but I wanted to request adding the distribution to `stats`.

There are some other skewed distributions not in `stats` in the same Wikipedia article that may be of use to SciPy users.

### PR DESCRIPTION
**Pull request by adivijaykumar**
https://cran.r-project.org/web/packages/sgt/vignettes/sgt.pdf and https://en.wikipedia.org/wiki/Skewed_generalized_t_distribution#skewed_Cauchy_distribution gives the expression for skewed Cauchy Distribution. Will add tests in due time. Fixes #7437 

As this in my first ENH fix, please let me know if I am on the right path.

### CODE REVIEW

**Comment by ev-br** [comment]
Yes, this is a very good start!

Several things to do still:

- [ ] Add this distribution to the generic test loop: add an entry to https://github.com/scipy/scipy/blob/master/scipy/stats/_distr_params.py

- [ ] Testing some values against R would be nice, see how it's done in e.g. `test_multivariate.py` (grep for "R code")

- [ ] An explicit form of cdf, possibly also ppf and/or rvs. These seem to be easy to add, and not having them makes a distributions slow.

**Comment by adivijaykumar** [comment]
http://dx.doi.org/10.1080/03610918.2014.1002617
This gives an algorithm for simulating truncated skewed Cauchy random variables. This is not what we want, is it? I didnt find any references for usual skewed Cauchy random variates. Or am I missing something here.

I can add the pdf and ppf, but the original paper does not give any expression for these. I can integrate the functions myself and derive explicit expressions, though. Shall I proceed?

**Comment by ev-br** [comment]
Yes, doing an integral is straighforward, seems like an easy way to proceed.

**Comment by adivijaykumar** [comment]
Added the ppf and cdf; please crosscheck.

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -4609,6 +4609,51 @@ def _entropy(self):
...
+    `skewcauchy` takes ``a`` as a skewness parameter
+    When a=0 the distribution is identical to the usual Cauchy distribution.
+
+    %(after_notes)s
+
+    %(example)s
+    """
+
+    def _argcheck(self, a):
+        return np.isfinite(a)
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: What happens at a=1 and -1 ? The pdf has an issue for x negative (positive).

**Review by ev-br — CHANGES_REQUESTED** [review]
Looks reasonable, but needs some more work. A test comparing a=0 and the Cauchy distribution would be good to have. Also a test against R.

File: scipy/stats/_distr_params.py [hunk 2]
```diff
@@ -91,6 +91,7 @@
     ['reciprocal', (0.0062309367010521255, 1.0062309367010522)],
     ['rice', (0.7749725210111873,)],
     ['semicircular', ()],
+    ['skewcauchy', (4.025168762123,)]
```
Comment (adivijaykumar) [inline scipy/stats/_distr_params.py]: Very sorry about that. Was in a rush, hence this.

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (adivijaykumar) [inline scipy/stats/_continuous_distns.py]: Nothing was explicitly mentioned in the sources, but it does seems to me that the distribution should be undefined for `np.abs(a) = 1`. Will make that correction.

**Comment by adivijaykumar** [comment]
Added tests against cauchy. Will figure out the R Tests soon.

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -4609,6 +4609,47 @@ def _entropy(self):
...
+    `skewcauchy` takes ``a`` as a skewness parameter
+    When a=0 the distribution is identical to the usual Cauchy distribution.
+
+    %(after_notes)s
+
+    %(example)s
+    """
+
+    def _argcheck(self, a):
+        return np.isfinite(a) and np.abs(a) != 1
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: https://en.m.wikipedia.org/wiki/Skewed_generalized_t_distribution

states that a (called lambda in that wiki page) is restricted to -1 < a < 1

**Comment by adivijaykumar** [comment]
That is right, I shall make the correction over the weekend.

**Comment by adivijaykumar** [comment]
But this also means that, the implementation of skewed normal is incorrect,
because even that does not impose this condition.
And again, following
https://en.wikipedia.org/wiki/Skewed_generalized_t_distribution#skewed_normal_distribution,
it ideally should

**Comment by hiccup7** [comment]
@adivijaykumar - You are very close to achieving success!  I greatly appreciate your efforts on this task.

CC: @ev-br

**Comment by zoj613** [comment]
ping @adivijaykumar 
did you lose interest on this PR?

**Comment by adivijaykumar** [comment]
Hey @zoj613, not really, I had forgotten about this. If you want to go ahead with this, please do!

File: scipy/stats/_distr_params.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_distr_params.py]: This was fixed.

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: as @ev-br mentioned later, the skew parameter is restricted to -1 < a < 1; this is fixed.

File: scipy/stats/_continuous_distns.py (hunk 3 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: Done.

**Comment by mdhaber** [comment]
@ev-br I picked this up, addressing all of your remaining comments, corrected the `ppf` implementation, and added a tutorial. Hope you don't mind, but I treated the comments as a checklist, resolving them as I went.

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -2134,6 +2134,53 @@ def test_expect2(self):
                         sum(v**2 * w for v, w in zip(y, py)), atol=1e-14)
 
 
+class TestSkewCauchy(object):
+    def setup_method(self):
+        np.random.seed(1234)
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py:2139]: Unused?

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -2134,6 +2134,53 @@ def test_expect2(self):
...
+        #        -4.797816025596743, 3.32619845547938, 2.7815675094985046,
+        #        3.700121482468191)
+        # pdf = dsgt(x, mu=0, lambda=lambda, sigma=1, q=1/2, mean.cent=FALSE,
+        #            var.adj = sqrt(2))
+        # cdf = psgt(x, mu=0, lambda=lambda, sigma=1, q=1/2, mean.cent=FALSE,
+        #            var.adj = sqrt(2))
+        # qsgt(cdf, mu=0, lambda=lambda, sigma=1, q=1/2, mean.cent=FALSE,
+        #      var.adj = sqrt(2))
+
+        np.random.seed(0)
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py:2168]: Duplicate seeding, here and in setup_method, one of them may go?

File: scipy/stats/tests/test_distributions.py [hunk 6]
```diff
@@ -2134,6 +2134,53 @@ def test_expect2(self):
...
+        a = np.random.rand(10) * 2 - 1
+        x = np.random.rand(10) * 10 - 5
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py:2170]: Maybe comment on relation between `x` in R code in comments and this one?

**Review by ev-br — APPROVED** [review]
Looks good modulo minor nits, which can be addressed in a follow-up.

**Comment by mdhaber** [comment]
Thanks! I'll address those this morning.

**Comment by ev-br** [comment]
Matt, would be able to squash the commits? 12 commits is a bit too much, and this PR cannot be squash-merged without losing authorship. I'd suggest making it two commits, one by you, the other by the OP. Or if you prefer it more fine-grained, fine by me too.

**Comment by mdhaber** [comment]
I could squash my commits, but I'm not sure about the best way to squash the original ones. I merged master in to avoid force-pushing to the original author's branch, so that puts those commits way back in the list. If I were to do an interactive rebase from here, I'd have to go some ~7000 commits back. I can think of at least one way to do it, but I am wondering what you would do?

File: scipy/stats/_continuous_distns.py [hunk 7]
```diff
@@ -7011,6 +7011,65 @@ def _entropy(self):
 semicircular = semicircular_gen(a=-1.0, b=1.0, name="semicircular")
 
 
+class skew_cauchy_gen(rv_continuous):
+    """A skewed Cauchy random variable.
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:7015]:
```suggestion
    r"""A skewed Cauchy random variable.
```

File: doc/source/tutorial/stats/continuous_skewcauchy.rst [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+This distribution is a generalization of the Cauchy distribution. It
+has a single shape parameter :math:`-1 < a < 1` that skews the distribution.
+The special case :math:`a=0` yields the Cauchy distribution.
+
+Functions
+---------
+
+.. math::
+   :nowrap:
+    \frac{1 - a}{2} + \frac{1 - a}{\pi} \atan\left(\frac{x}{1 - a} \right)
```
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_skewcauchy.rst:15]:
```suggestion

```

File: doc/source/tutorial/stats/continuous_skewcauchy.rst [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    \begin{eqnarray*}
+    f(x, a) & = & \frac{1}{\pi \left(\frac{x^2}{\left(a x + 1 \right)^2} + 1 \right)},\quad x\ge0; \\
+                 & = & \frac{1}{\pi \left(\frac{x^2}{\left(-a x + 1 \right)^2} + 1 \right)},\quad x<0.
+    F(x, a) & = & \frac{1 - a}{2} + \frac{1 + a}{\pi} \arctan\left(\frac{x}{1 + a} \right),\quad x\ge0; \\
+                 & = & \frac{1 - a}{2} + \frac{1 - a}{\pi} \arctan\left(\frac{x}{1 - a} \right),\quad x<0. \\
```
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_skewcauchy.rst:19]:
```suggestion
    f(x, a) & = & \frac{1}{\pi \left(\frac{x^2}{\left(a x + 1 \right)^2} + 1 \right)},\quad x\ge0; \\
                 & = & \frac{1}{\pi \left(\frac{x^2}{\left(-a x + 1 \right)^2} + 1 \right)},\quad x<0. \\
    F(x, a) & = & \frac{1 - a}{2} + \frac{1 + a}{\pi} \arctan\left(\frac{x}{1 + a} \right),\quad x\ge0; \\
                 & = & \frac{1 - a}{2} + \frac{1 - a}{\pi} \arctan\left(\frac{x}{1 - a} \right),\quad x<0.
```

**Comment by ev-br** [comment]
If the interactive rebase is challenging, I'd just squash into three commits: all from 2017, merge the master, and all new developments of 2020 and 2021.

**Comment by ev-br** [comment]
And for your cherry-picking approach, I'd squash  into three anyway, so there's only two commits to cherry-pick, instead of 15 :-)

**Comment by mdhaber** [comment]
I'm not aware of being able to squash except via interactive rebase. If I do an interactive rebase, there is no need to keep the merge commit. 
The problem is that I don't want to do an interactive rebase involving ~7000 commits that have occurred since the first of the original author:s commits.

File: doc/source/tutorial/stats/continuous_skewcauchy.rst [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+This distribution is a generalization of the Cauchy distribution. It
+has a single shape parameter :math:`-1 < a < 1` that skews the distribution.
+The special case :math:`a=0` yields the Cauchy distribution.
+
+Functions
+---------
+
+.. math::
+   :nowrap:
+    \begin{eqnarray*}
```
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_skewcauchy.rst:15]:
```suggestion

    \begin{eqnarray*}
```

**Comment by mdhaber** [comment]
That was unpleasant.
I created gh-13374 to avoid force pushing to @adivijaykumar's repo. Preserving this also makes it possible to confirm that there is no difference between the new PR and the rebased version.

**Pull request closed without merging**